### PR TITLE
Fix deprecated issue/pull number context

### PR DESCRIPTION
**Pull request by mheap**
**Why?**

The `number` parameter for Issues and Pull Requests has been deprecated for a while. The only mention I can find of it is in `octokit/routes`: https://github.com/octokit/routes/commit/90f232a

Up until this point it's been working fine, but I was building an Action yesterday when I received a `404`. I worked around it with the following code, but the experience isn't as good:

```javascript
await tools.github.issues.removeLabel({
  ...tools.context.repo,
  issue_number: tools.context.issue.number,
  name
});
```

**How?**

Check the provided payload for the existance of certain keys. If `.issue` exists, then return `issue_number`. If `.pull_request` exists, return `pull_number` otherwise fall back to the original behaviour reading `.number` from the payload root.

---

- [x] Tests have been added/updated (if necessary)
- [x] Documentation has been updated (if necessary)

### CODE REVIEW

File: src/context.ts [hunk 1]
```diff
@@ -73,10 +73,22 @@ export class Context {
...
+      // If it's an issue, the API expects issue_number
+      context.issue_number = payload.issue.number
+    } else if (payload.pull_request) {
+      // If it's a PR, the API expects pull_number
+      context.pull_number = payload.pull_request.number
+    } else {
+      // Otherwise default to the old behaviour
+      context.number = payload.number
+    }
+
```
Comment (JasonEtco) [inline src/context.ts:90]: I think we should add this for backwards compat (with a comment explaining that):

```js
context.number = context.number || context.issue_number || context.pull_number
```

Otherwise this'll be a breaking change and require a major version bump right now, and we can ease into a bit - and get rid of this in the next major version.

File: src/context.ts [hunk 2]
```diff
@@ -73,10 +73,22 @@ export class Context {
   public get issue () {
     const payload = this.payload
 
-    return {
-      ...this.repo,
-      number: (payload.issue || payload.pull_request || payload).number
+    const context: { [k: string]: any } = {
```
Comment (JasonEtco) [inline src/context.ts:76]: The variable name `context` is confusing here, because its in the `Context` class - I originally read it as an instance of `Context`. How about:

* issue
* data

**Comment by mheap** [comment]
@JasonEtco I've been thinking about this and I'm not sure it's the right fix. It'd work for common use cases, but if you're trying to add a label to a `pull_request` it's going to give you the wrong payload.

Would it make sense to split `issue()` in to `issue()` and `pullRequest()` and let the consumer explicitly choose which they prefer?

**Comment by JasonEtco** [comment]
> Would it make sense to split issue() in to issue() and pullRequest() and let the consumer explicitly choose which they prefer?

I see what you're saying - I'd opt to have two properties:

```ts
class Context {
  get issue_number () { return this.payload.issue?.number }
  get pull_number () { return this.payload.pull_request?.number }
}
```

These would be simple getters that look for `payload.(issue|pull_request).number`. Then you'd use them _with_ `context.repo`:

```js
{
  ...context.repo,
  issue_number: context.issue_number
}
```

It's not as clean as have just one getter, but since the `issue_number/pull_number` keys are different I don't think there's a way to only have one. What do you think @mheap? The alternative, like you said, would be:

```ts
class Context {
  get issue () { return { ..this.repo, this.payload.issue?.number } }
  get pullRequest () { return { ...this.repo, this.payload.pull_request?.number } }
}
```

I'm not sure which I prefer 🤔

**Comment by mheap** [comment]
tl;dr: Have two getters, `issue` and `pullRequest`. Also keep the fallback implementation already proposed to allow `pullRequest` to return `{issue_number: payload.pull_request.number}`

## More details

I can see two issues here:

### Where to pull the number from

This is covered by the PR in it's current state

### The format that we return the data in

I'd like to be able to do `...context.issue` or `...context.pullRequest` depending on if we're using an Issue or a PR endpoint.

`...context.issue` should use the implementation in this PR to extract `number` and return an `issue_number` key in the context. The value of `number` *may* come from `payload.pull_request.number` but it needs to be returned as `issue_number` if we're interacting with Issue features e.g. Labels

**Comment by mheap** [comment]
@JasonEtco Any thoughts on the above?

**Comment by JasonEtco** [comment]
@mheap that makes sense to me!

**Comment by JasonEtco** [comment]
@mheap I'm working on v5 that includes a bunch of breaking changes. I'm going to hold off until this PR is in (take your time though, I'm not rushing to release v5) so that we can nicely document this as part of the release.
